### ISSUE TITLE
java.lang.IllegalStateException: Deque is too big

### ISSUE THREAD

**Issue by shettigarsanthosh** [InsertKoinIO/koin#1353]
Facing following issue, when tried to use access get() in KoinComponent Scope

java.lang.IllegalStateException: Deque is too big.
at kotlin.collections.ArrayDeque.ensureCapacity(ArrayDeque.java:59)
at kotlin.collections.ArrayDeque.addFirst(ArrayDeque.java:124)
at org.koin.core.scope.Scope.resolveInstance(Scope.java:238)
at org.koin.core.scope.Scope.get(Scope.java:204)

Can you please help to understand the possible root cause for above crash ?
Koin version - 3.1.5

**Comment by stuebingerb** [comment]
I would like to jump in here as we have been experiencing the same exception with Koin 3.1.6 and now with Koin 3.2.0.

The issue seems to be that the `_parameterStack` is filling up. I am not (yet) suggesting that there is a bug in Koin but we have a rather complex Koin setup and would also be happy to get some starting points for further investigation. We can also see that now all of our pods are affected equally, so there seem to be some yet-to-be-found special circumstances.

If it helps, we have created a metric to observe the `_parameterStack.size` for the root scope (which is our only scope).

General overview over 2 days, with each line representing one of our pods, showing sizes of up to 17.000 elements:
![image](https://user-images.githubusercontent.com/41049452/173061747-1a9bab92-cb32-42f3-8b95-31fd3ed6be1a.png)

And a single failing pod, showing the size going down to -2, presumably due to integer overflow:
![image](https://user-images.githubusercontent.com/41049452/173062172-e6ed572b-2278-4977-8808-0b603e708384.png)

**Comment by shettigarsanthosh** [comment]
Any update root cause for this issue ?

**Comment by arnaudgiuliani** [comment]
Clearly need investigation yes 👍

**Comment by stuebingerb** [comment]
Can we do anything to support?

**Comment by stuebingerb** [comment]
As of now I don't have any sample code that I would be allowed to share unfortunately.

But I may have a lead: We have classes implementing `KoinComponent` which we are _not_ injecting via Koin but instead manually instantiate via their constructor - within the context of a Koin managed class. Simplified:

```
module {
    single { Service() }
}
```
```
class HelperB : KoinComponent {
    fun bar() { /* Implementation */ }
}
class HelperA : KoinComponent {
    fun foo() {
        HelperB().bar()
    }
}
class Service : KoinComponent {
    fun process() {
        HelperA().foo()
    }
}
class Wrapper : KoinComponent {
    get<Service>.process()
}
```

And when I look at the parameterStack at sizes > 0 I can see `ParametersHolder` instances (only) for `HelperA` and `HelperB` (along with some `null` entries).

Now when I add `HelperA` to the module definition and replace the instance creation with `get<HelperA>()` it disappears from the parameter stack and only instances for `HelperB` remain:

```
module {
    single { Service() }
    single { HelperA() }
}
```
```
class HelperB : KoinComponent {
    fun bar() { /* Implementation */ }
}
class HelperA : KoinComponent {
    fun foo() {
        HelperB().bar()
    }
}
class Service : KoinComponent {
    fun process() {
        get<HelperA>().foo()
    }
}
class Wrapper : KoinComponent {
    get<Service>.process()
}
```

I still don't quite understand why the same code seems to behave so differently in different pods (and why the parameter size still goes _down_ sometimes) but to me this looks like a very potential root cause candidate (and incorrect use of Koin, thus qualifying indeed as "not a bug").

Does any of that sound plausible?

**Comment by stuebingerb** [comment]
I think I can confirm that the above change (getting rid of manual instance creation within a Koin managed context) fixed the issue for us. Maybe @shettigarsanthosh can confirm for them, otherwise I'd leave it up to you how to proceed here.

**Comment by PenzK** [comment]
Also faced such issue after migrating to 3.2.1 from 2.1.5

We use dynamic resolving by named argument
We have content provider which receives a named argument and some params
`getKoin().getOrNull<IContentProviderDelegate>(named(contentType.name)) {
            parametersOf(params ?: "")
        }`
        
        
on other end we have 

` factory<IContentProviderDelegate>(named(ContentType.name)) { (request: String) ->
        SomeProviderDelegate(
            jsonRequest = request,
            someStaff = get(),
           ...
        )
    }`
    
   
 For now assuming that due to concurrent calls to content provider arraydeque  params logic has some issues
 
 In some cases we got `ArrayDeque is empty `, in some `ArrayDeque is too big`

**Comment by partho-mt** [comment]
I'm facing this issue while using `inject` as well. My code is as follows:
`val service: Service = injectByComponent { parametersOf(param) }`

And my `injectByComponent` method is:
```
inline fun <reified T> injectByComponent(
    noinline parameters: ParametersDefinition? = null
): T {
    return object : KoinComponent {
        val value: T by inject(parameters = parameters)
    }.value
}
```

Koin version: 3.2.2

**Comment by poldz123** [comment]
Any update for this issue? as we are having the same crash as well

**Comment by arnaudgiuliani** [comment]
are you all using it in ktor?

**Comment by partho-mt** [comment]
@arnaudgiuliani I am.

**Comment by PenzK** [comment]
> are you all using it in ktor?

We aren't

**Comment by arnaudgiuliani** [comment]
I believe we can try this fix: b2efb2248abde21916f666415f0d4c0f82244a59

this is pushing synchronization when we have parameters on stack. This would avoid having dangling parameters on stack

I will publish it in Koin 3.3.

**Comment by arnaudgiuliani** [comment]
Feel free to reopen an issue if you need 👍
